### PR TITLE
Handle camera movement when the clicked position is void

### PR DESCRIPTION
**Pull request by NEDJIMAbelgacem**
## Description
When the clicked area has nothing rendered (a void area), the camera behaves weirdly as it picks a very far point as a rotation center or a drag point. This PR disables the camera movement when that picked point is void.

### CODE REVIEW

**Comment by nyalldawson** [comment]
How does this work if you only have a single point cloud layer loaded? Does it mean you have to click right on a point to navigate?

**Comment by uclaros** [comment]
Nyall's concern is valid also for cases where the background if the pointcloud is void, like a front view.
My suggestion would be to implement the following logic:
- if depth is 1, query the depth buffer's neighboring 8 pixels and use the minimum value.
- if still 1, go for 15 or 24 (4x4 or 5x5 neighbours).
- if still 1, it means the user has probably clicked straight into the void. In that case instead of ignoring his input, calculate the average value of the whole depth buffer and use this to get the center of rotation.

**Comment by NEDJIMAbelgacem** [comment]
Thank you for the feedback @uclaros & @nyalldawson
I changed the behaviour to be like this:
- Query the 9 by 9 square around the clikced pixel for depth
- If the depth is 1 (meaning we are in a non rendered area), we calculate the average of the rendered pixels (depths that are different than 1)

File: src/3d/qgscameracontroller.cpp [hunk 1]
```diff
@@ -27,6 +27,7 @@
 #include <Qt3DRender/QObjectPicker>
 #include <Qt3DRender/QPickEvent>
 #include <Qt3DInput>
+#include <QElapsedTimer>
```
Comment (uclaros) [inline src/3d/qgscameracontroller.cpp:30]:
```suggestion
```

**Review by uclaros — CHANGES_REQUESTED** [review]
In general this gives great usability! My only concern is with the right mouse zooming, where zooming to void was not troublesome..

File: src/3d/qgscameracontroller.cpp [hunk 2]
```diff
@@ -494,7 +546,12 @@ void QgsCameraController::onPositionChangedTerrainNavigation( Qt3DInput::QMouseE
 
     if ( !mDragPointCalculated )
     {
-      double depth = Qgs3DUtils::decodeDepth( mDepthBufferImage.pixelColor( mDragButtonClickPos.x(), mDragButtonClickPos.y() ) );
+      double depth = sampleDepthBuffer( mDepthBufferImage, mDragButtonClickPos.x(), mDragButtonClickPos.y() );
+      if ( depth >= 1 )
+      {
+        depth = calculateAverageDepth( mDepthBufferImage );
+      }
```
[the remark this answers is not in the document]
Comment (NEDJIMAbelgacem) [reply src/3d/qgscameracontroller.cpp:553]: The Panning and rotating are totally unrealted to the right button zooming.
I think this allows for more uniform behaviour accross navigation controls and this change still allows precise zooming with the right button. In any case you wouldn't want to zoom towards a very far point anyway...

File: src/3d/qgscameracontroller.cpp [hunk 3]
```diff
@@ -308,6 +309,45 @@ double QgsCameraController::cameraCenterElevation()
...
+double QgsCameraController::calculateAverageDepth( const QImage &buffer )
+{
+  double depth = 0;
+  int samplesCount = 0;
+  for ( int x = 0; x < mDepthBufferImage.width(); ++x )
+  {
+    for ( int y = 0; y < mDepthBufferImage.height(); ++y )
+    {
+      const QRgb &pixel = buffer.pixel( x, y );
+      double d = ( ( qRed( pixel ) / 255.0 + qGreen( pixel ) ) / 255.0 + qBlue( pixel ) ) / 255.0;
```
[the remark this answers is not in the document]
Comment (NEDJIMAbelgacem) [reply src/3d/qgscameracontroller.cpp:339]: This is faster by around 50% in debug :)

File: src/3d/qgscameracontroller.h [hunk 4]
```diff
@@ -277,6 +277,9 @@ class _3D_EXPORT QgsCameraController : public Qt3DCore::QEntity
 
     double cameraCenterElevation();
 
+    // Returns the minimum depth value in the square [px - 3, px + 3] * [py - 3, py + 3]
```
Comment (uclaros) [inline src/3d/qgscameracontroller.h:280]:
```suggestion
    /**
     * Returns the minimum depth value in the square [px - 3, px + 3] * [py - 3, py + 3]
     * If the value is 1, the average depth of all non void pixels is returned instead.
     */
```

File: src/3d/qgscameracontroller.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (uclaros) [reply src/3d/qgscameracontroller.cpp:338]: I would prefer if you always used the same function for depth decoding, so maybe change `Qgs3DUtils::decodeDepth` to use `pixel()` instead.

**Review by uclaros — CHANGES_REQUESTED** [review]
Minor stuff again.

**Review by uclaros — APPROVED** [review]
Yay! \o/

**Pull request merged**